### PR TITLE
Intermediate secret && configMap update 

### LINKED ISSUE #210: The operator does not reconcile (do not watch) on the intermediate secret
When a Backing service is bound with an Application by the operator an intermediate secret is used to store all the binding values that are injected into the Application.

When this intermediate secret's data is modified the operator is expected to detect this, reconcile and revert the intermediate secret back.

That is actually not happening - when the intermediate secret is modified, the operator ignores it.

Possible cause of the e2e failures described in https://jira.coreos.com/browse/APPSVC-147

### PR DESCRIPTION
**Pull request by DhritiShikhar**
fixes #210

### CODE REVIEW

**Comment by DhritiShikhar** [comment]
@sbose78 Can you please review this PR?

File: pkg/controller/servicebindingrequest/sbrcontroller.go [hunk 1]
```diff
@@ -32,6 +32,9 @@ var (
 	// defaultPredicate default predicate functions
 	defaultPredicate = predicate.Funcs{
 		UpdateFunc: func(e event.UpdateEvent) bool {
+			if e.ObjectNew.GetObjectKind().GroupVersionKind().Kind == "secret" {
+				return false
```
Comment (sbose78) [inline pkg/controller/servicebindingrequest/sbrcontroller.go]: Should be `true` right?

Comment (DhritiShikhar) [reply pkg/controller/servicebindingrequest/sbrcontroller.go]: @sbose78 Yes. I have made a couple of more changes. Will push those changes in some time.

File: pkg/controller/servicebindingrequest/sbrcontroller.go [hunk 2]
```diff
@@ -32,6 +32,9 @@ var (
 	// defaultPredicate default predicate functions
 	defaultPredicate = predicate.Funcs{
 		UpdateFunc: func(e event.UpdateEvent) bool {
+			if e.ObjectNew.GetObjectKind().GroupVersionKind().Kind == "secret" {
```
Comment (pmacik) [inline pkg/controller/servicebindingrequest/sbrcontroller.go]: @DhritiShikhar the `Kind` might actually be `Secret`(with capital `S`) so I'd suggest to use the one string comparison ignoring case, so:
```go
if strings.EqualFold(e.ObjectNew.GetObjectKind().GroupVersionKind().Kind, "secret") {
   return true
}
```

Comment (pmacik) [reply pkg/controller/servicebindingrequest/sbrcontroller.go]: like this https://github.com/DhritiShikhar/service-binding-operator/pull/1 :)

**Comment by pmacik** [comment]
@sbose78 @DhritiShikhar @Avni-Sharma I just have an idea:
What if we compute some kind of checksum (hash) from the `Data` and save it in annotations (kind of sign it) and then compute/validate the checksum for each update event and compare it with the one in annotations and then reconcile only if the signatures do not match?

**Comment by DhritiShikhar** [comment]
@pmacik Okay. But shouldn't that be done for every object including `secret`?

**Comment by DhritiShikhar** [comment]
Output:


```
>>>>>>>>>>>>>>>>> Secret
OLD_DATA************* &{map[apiVersion:v1 data:map[DATABASE_CONFIGMAP_DB_HOST:MTcyLjMwLjgxLjE2MQ== DATABASE_CONFIGMAP_DB_NAME:ZGItZGVtbw== DATABASE_CONFIGMAP_DB_PASSWORD:cGFzc3dvcmQ= DATABASE_CONFIGMAP_DB_PORT:NTQzMg== DATABASE_CONFIGMAP_DB_USERNAME:cG9zdGdyZXM= DATABASE_CONFIGMAP_PASSWORD:cGFzc3dvcmQ= DATABASE_CONFIGMAP_USER:cG9zdGdyZXM= DATABASE_DBCONNECTIONIP:MTcyLjMwLjE3NC44Mw== DATABASE_DBCONNECTIONPORT:NTQzMg== DATABASE_DBNAME:ZGItZGVtbw== DATABASE_SECRET_PASSWORD:cGFzc3dvcmQ= DATABASE_SECRET_USER:cG9zdGdyZXM=] kind:Secret metadata:map[creationTimestamp:2019-11-01T12:00:46Z name:binding-request namespace:service-binding-demo resourceVersion:109695 selfLink:/api/v1/namespaces/service-binding-demo/secrets/binding-request uid:12ade816-94f1-41f1-99bd-dd2684d71a98] type:Opaque]}
NEW_DATA************* &{map[apiVersion:v1 data:map[DATABASE_CONFIGMAP_DB_HOST:MTcyLjMwLjgxLjE2MQ== DATABASE_CONFIGMAP_DB_NAME:ZGItZGVtbw== DATABASE_CONFIGMAP_DB_PASSWORD:cGFzc3dvcmQ= DATABASE_CONFIGMAP_DB_PORT:NTQzMg== DATABASE_CONFIGMAP_DB_USERNAME:cG9zdGdyZXM= DATABASE_CONFIGMAP_PASSWORD:cGFzc3dvcmQ= DATABASE_CONFIGMAP_USER:cG9zdGdyZXM= DATABASE_DBCONNECTIONIP:MTcyLjMwLjE3NC44Mw== DATABASE_DBCONNECTIONPORT:NTQzMg== DATABASE_DBNAME:ZGItZGVtbw== DATABASE_SECRET_PASSWORD:cGFzc3dvcmQ= DATABASE_SECRET_USER:cG9zdGdyZXM=] kind:Secret metadata:map[annotations:map[service-binding-operator.apps.openshift.io/binding-name:binding-request service-binding-operator.apps.openshift.io/binding-namespace:service-binding-demo] creationTimestamp:2019-11-01T12:00:46Z name:binding-request namespace:service-binding-demo resourceVersion:109715 selfLink:/api/v1/namespaces/service-binding-demo/secrets/binding-request uid:12ade816-94f1-41f1-99bd-dd2684d71a98] type:Opaque]}
2019-11-01T18:32:52.192+0530	INFO	reconciler	All done!	{"Request.Namespace": "service-binding-demo", "Request.Name": "binding-request", "ServiceBindingRequest.Name": "binding-request"}
2019-11-01T18:32:52.192+0530	INFO	reconciler	Reconciling ServiceBindingRequest...	{"Request.Namespace": "service-binding-demo", "Request.Name": "binding-request"}
2019-11-01T18:32:54.296+0530	INFO	reconciler	Binding applications with intermediary secret.	{"Request.Namespace": "service-binding-demo", "Request.Name": "binding-request", "ServiceBindingRequest.Name": "binding-request"}
```

**Comment by pmacik** [comment]
> @pmacik Okay. But shouldn't that be done for every object including `secret`?

I think so, yes - ideally.

But that would require extending the annotation mechanism with the signature and validation, and also with the ways of computing the hash per object kind.

**Comment by DhritiShikhar** [comment]
@pmacik Then we can address that in another PR. We should create a task for it.

**Comment by DhritiShikhar** [comment]
Discussion from Friday standup:
[1] Add the same check for ConfigMap

**Comment by pmacik** [comment]
/lgtm
The e2e tests fail for another reason.

**Comment by DhritiShikhar** [comment]
@sbose78 Can you merge this PR?

**Comment by DhritiShikhar** [comment]
@sbose78 I cannot merge this PR. Can you manually merge this PR?

**Pull request merged**